body/stretch_body/pimu.py

```diff
@@ -488,3 +473,114 @@ def step_sentry(self):
+            self.status['voltage']=self.get_voltage(unpack_float_t(s[sidx:]));sidx+=4
+            self.status['current'] = self.get_current(unpack_float_t(s[sidx:]));sidx+=4
+            self.status['temp'] = self.get_temp(unpack_float_t(s[sidx:]));sidx+=4
+
+            for i in range(4):
+                self.status['cliff_range'][i]=unpack_float_t(s[sidx:])
+                sidx+=4
+
+            self.status['state'] = unpack_uint32_t(s[sidx:])
+            sidx += 4
+
+            self.status['at_cliff']=[]
+            self.status['at_cliff'].append((self.status['state'] & PimuBase.STATE_AT_CLIFF_0) != 0)
+            self.status['at_cliff'].append((self.status['state'] & self.STATE_AT_CLIFF_1) != 0)
+            self.status['at_cliff'].append((self.status['state'] & self.STATE_AT_CLIFF_2) != 0)
+            self.status['at_cliff'].append((self.status['state'] & self.STATE_AT_CLIFF_3) != 0)
+            self.status['runstop_event'] = (self.status['state'] & self.STATE_RUNSTOP_EVENT) != 0
+            self.status['cliff_event'] = (self.status['state'] & self.STATE_CLIFF_EVENT) != 0
+            self.status['fan_on'] = (self.status['state'] & self.STATE_FAN_ON) != 0
+            self.status['buzzer_on'] = (self.status['state'] & self.STATE_BUZZER_ON) != 0
+            self.status['low_voltage_alert'] = (self.status['state'] & self.STATE_LOW_VOLTAGE_ALERT) != 0
+            self.status['high_current_alert'] = (self.status['state'] & self.STATE_HIGH_CURRENT_ALERT) != 0
+            self.status['over_tilt_alert'] = (self.status['state'] & self.STATE_OVER_TILT_ALERT) != 0
+
+            self.status['timestamp'] = self.timestamp.set(unpack_uint64_t(s[sidx:])); sidx += 8
+            self.imu.status['timestamp'] = self.status['timestamp']
+            self.status['bump_event_cnt'] = unpack_uint16_t(s[sidx:]);sidx += 2
+            self.status['debug'] = unpack_float_t(s[sidx:]); sidx += 4
+            self.status['cpu_temp']=self.get_cpu_temp()
+            return sidx
+
+# ######################## PIMU #################################
+class Pimu(PimuBase):
+    """
+    API to the Stretch RE1 Power and IMU board (Pimu)
+    """
+    def __init__(self, event_reset=False):
+        PimuBase.__init__(self, event_reset)
+        self.supported_protocols = {'p0': Pimu_Protocol_P0, 'p1': Pimu_Protocol_P1}
+
+    def startup(self):
+        """
+        First determine which protocol version the uC firmware is running.
+        Based on that version, replaces PimuBase class inheritance with a inheritance to a child class of PimuBase that supports that protocol
+        """
+        PimuBase.startup(self)
+        if self.hw_valid:
+            if self.board_info['protocol_version'] in self.supported_protocols:
+                Pimu.__bases__ = (self.supported_protocols[self.board_info['protocol_version']],)
+                self.imu.__bases__= (self.imu.supported_protocols[self.board_info['protocol_version']],)
```

Comment:
Should be `IMU.__bases__` instead of `self.imu.__bases__`. Fails with NotImplementedError on the latter.

Reply:
Fixed